include/quadrilateral_gimp_element.tcc

```diff
@@ -0,0 +1,178 @@
+      double ni = local_nodes(n, i);
+      double npni = xi(i) - ni;  // local particle  - local node
+      //! Conditional shape function statement see: Bardenhagen 2004
+      if (npni <= (-element_length - particle_size(i))) {
+        sni(i) = 0.;
+        dni(i) = 0.;
+
+      } else if ((-element_length - particle_size(i)) < npni &&
+                 npni <= (-element_length + particle_size(i))) {
+
+        sni(i) = std::pow(element_length + particle_size(i) + npni, 2.) /
+                 (4. * (element_length * particle_size(i)));
+        dni(i) = (element_length + particle_size(i) + npni) /
+                 (2. * element_length * particle_size(i));
+      } else if ((-element_length + particle_size(i)) < npni &&
+                 npni <= -particle_size(i)) {
+        sni(i) = 1. + (npni / element_length);
+        dni(i) = 1. / element_length;
+      } else if (-particle_size(i) < npni && npni <= particle_size(i)) {
+        sni(i) = 1. - (((npni * npni) + (particle_size(i) * particle_size(i))) /
+                       (2. * element_length * particle_size(i)));
+        dni(i) = -(npni / (element_length * particle_size(i)));
+      } else if (particle_size(i) < npni &&
+                 npni <= element_length - particle_size(i)) {
+        sni(i) = 1. - (npni / element_length);
+        dni(i) = -(1. / element_length);
+      } else if ((element_length - particle_size(i)) < npni &&
+                 npni <= element_length + particle_size(i)) {
+        sni(i) = std::pow(element_length + particle_size(i) - npni, 2.) /
+                 (4. * element_length * particle_size(i));
+        dni(i) = -((element_length + particle_size(i) - npni) /
+                   (2. * element_length * particle_size(i)));
+      } else if ((element_length + particle_size(i)) < npni) {
+        sni(i) = 0.;
+        dni(i) = 0.;
+      }
+    }
+    grad_shapefn(n, 0) = dni(0) * sni(1);
+    grad_shapefn(n, 1) = dni(1) * sni(0);
+  }
+  return grad_shapefn;
+}
+
+//! Return the B-matrix of a Quadrilateral Element at a given local
+//! coordinate for a real cell
+template <unsigned Tdim, unsigned Tnfunctions>
+inline std::vector<Eigen::MatrixXd>
+    mpm::QuadrilateralGIMPElement<Tdim, Tnfunctions>::bmatrix(
+        const VectorDim& xi, const Eigen::MatrixXd& nodal_coordinates,
+        const VectorDim& particle_size,
+        const VectorDim& deformation_gradient) const {
+  return this->mpm::QuadrilateralElement<Tdim, Tnfunctions>::bmatrix(
+      xi, nodal_coordinates);
+}
+
+//! Compute Jacobian with particle size and deformation gradient
+template <unsigned Tdim, unsigned Tnfunctions>
+inline Eigen::Matrix<double, Tdim, Tdim>
+    mpm::QuadrilateralGIMPElement<Tdim, Tnfunctions>::jacobian(
+        const VectorDim& xi, const Eigen::MatrixXd& nodal_coordinates,
+        const VectorDim& particle_size,
+        const VectorDim& deformation_gradient) const {
+  return this->mpm::QuadrilateralElement<Tdim, Tnfunctions>::jacobian(
```

Comment:
This won't work Tdim, Tnfunctions won't match. You need to copy and paste the content here